app/featuresmodel.cpp

```diff
@@ -25,31 +32,74 @@ void FeaturesModel::populate()
 {
   if ( mLayer )
   {
+    mFetchingResults = true;
+    emit fetchingResultsChanged();
     beginResetModel();
     mFeatures.clear();
+    endResetModel();
 
     QgsFeatureRequest req;
     setupFeatureRequest( req );
 
-    QgsFeatureIterator it = mLayer->getFeatures( req );
-    QgsFeature f;
+    int searchId = mNextSearchId.fetchAndAddOrdered( 1 );
+    QgsVectorLayerFeatureSource *source = new QgsVectorLayerFeatureSource( mLayer );
+    mSearchResultWatcher.setFuture( QtConcurrent::run( &FeaturesModel::fetchFeatures, this, source, req, searchId ) );
+  }
+}
+
+QgsFeatureList FeaturesModel::fetchFeatures( QgsVectorLayerFeatureSource *source, QgsFeatureRequest req, int searchId )
+{
+  QElapsedTimer t;
+  t.start();
+  std::unique_ptr<QgsVectorLayerFeatureSource> fs( source );
+
+  QgsFeatureIterator it = fs->getFeatures( req );
+  QgsFeature f;
+  QgsFeatureList fl;
 
-    while ( it.nextFeature( f ) )
+  // a search might have been queued if no threads were available in the pool, so we also
+  // check if canceled before we start as the first iteration can potentially be slow
+  bool canceled = searchId + 1 != mNextSearchId.loadAcquire();
+  if ( canceled )
+    qDebug() << QString( "Search (%1) was cancelled before it started!" ).arg( searchId );
```

Comment:
a check for cancelled and an early exist could be actually done before the getFeatures() call? (to save time doing all the call initializing the search)